### ISSUE TITLE
Main character's side image pops in at (seemingly) random places

### ISSUE THREAD

**Issue by Tymedust** [renpy/renpy#4828]
Hi!

So, this is a strange one. I've reached out to other Ren'Py users far beyond my own skill level in terms of coding/programming, and so far no one seems to have any idea what's going on, so I'm hoping posting it here will finally resolve this strange issue.

The problem is basically that my MC's portrait, which is a side image, slides in when it shouldn't. It appears, stays in place very briefly, and then slides back out. This can even happen in scenes where she isn't part of the proceedings, such as when two other characters are talking and she isn't included at all. Stranger still is that this happens at random places for random people. No two people encounter the issue at the same place as each other, and I even have many who have played the whole game through without having it happen once. 

It's happened to me too, but I've chalked it up to like temporary coding hiccups or something, but now I realize that it's something that can happen to others as well. Quitting the game and loading a save resolves the issue again, leading me to believe that it can possibly be related to memory or something (but don't quote me on that because my coding skills are very limited).

One of my players managed to capture an image of it happening at a particular instance:

![SPOILER_image](https://github.com/renpy/renpy/assets/123891561/10746610-2d79-4647-9696-3b18b6a83ae3)

Here you can see the MC's image slide out (having just previously slid in), as another character starts talking in a scene where she has no dialogue. In code, it looks like this:

```
    rs "You are {i}such{/i} a {i}catch{/i}."
    hide vn
    show vs at pos22
    vs "..."
    $ renpy.pause(0.5, hard=True)  
    hide vs
    show vn at pos22
    vn "Thank you."
    vn "Well, if that's all..."
```

MC slides in after this line:

`    vs "..."`

And then slides out again when the following line starts:

`    vn "Thank you."`

So, she's basically occupying the space that the pause takes up. And as you can see, she's nowhere to be found in the code.

One of my players suggested this:

> If it helps, my hunch is that it has less to do with specific scenes and maybe with the interface while Isobel is on screen, maybe rollback, pressing tab to skip, or reloading save during dialogue? Not sure about others but whenever the cycling happens it's usually because I did one of those three actions recently

I *think* that might have been the case for me too when I've encountered it, and I'm fairly sure that I might have done so during a pause as well, so maybe the answer lies there somewhere? However, trying to purposefully recreate the issue by doing rollback or reload at the instance above (depicted in the screenshot) didn't work either. The bug seems to be a bit elusive, in other words.

As for how the side image is set up, I use a simple definition like so:

`define e = Character("Isobel", what_prefix='"', what_suffix='"', image="e", ctc="ctc_blink", ctc_position="nestled")`

And when the character is called in-game, it looks like this for example:

`    e s "And... what are my results?"`

The images themselves are named as "side e n", "side e s", etc where "e" is the character and the second letter is the expression.

I also use an animation to define how her image slides in and out, which looks like this:

```
define config.side_image_change_transform = change_transform

transform change_transform(old, new):
    contains:
        old
        yalign 1.0
        xpos 0.0 xanchor 0.0
        linear 0.2 xanchor 1.0
    contains:
        new
        yalign 1.0
        xpos 0.0 xanchor 1.0
        linear 0.2 xanchor 0.0
```

It might also be worth noting that I have seen this issue in my previous games as well (all using the same setup as seen above) but for some reason there seems to be more of it in this one, though perhaps that's just by chance.

Let me know if you need more information and I'll do my best to dig it up.

Thank you for your time and I hope you'll be able to help me in solving this conundrum. :)

**Comment by renpytom** [comment]
How are you defining the side images, both here and in the say screen? Pause shows the empty window, which is the same thing as the narrator speaking. So I'm wondering if the narrator speaking is what's causing the main character's side image to appear.

**Comment by Tymedust** [comment]
> How are you defining the side images, both here and in the say screen? Pause shows the empty window, which is the same thing as the narrator speaking. So I'm wondering if the narrator speaking is what's causing the main character's side image to appear.

The character definition is as follows:

`define e = Character("Isobel", what_prefix='"', what_suffix='"', image="e", ctc="ctc_blink", ctc_position="nestled")`

There is no separate side image definition other than that. I've simply named the files "side e n", "side e s" etc and then put them in my image folder. Then they are called as such:

`    e s "So... I'm done?"`

(In that case it shows the image "side e s" along with the text of what she's saying, being character "e" and using the image of "e")

The say screen simply looks like this, I haven't altered it:

```
screen say(who, what):
    style_prefix "say"

    window:
        id "window"

        if who is not None:

            window:
                id "namebox"
                style "namebox"
                text who id "who"

        text what id "what"
```

I have a defined narrator as well:

`define n = Character(ctc="ctc_blink", ctc_position="nestled")`

If I follow you correctly, though, the problem may be that Ren'Py during pause shows an empty box, which somehow triggers the side image to appear? Almost as if the engine believes my MC to be the narrator or something...? 

I have no idea how that would work though if so, that's way beyond my expertise. :(

**Comment by Tymedust** [comment]
Bumping this with some potentially useful information that's come to light recently.

I've continued investigating occurrences of the issue, and it seems like **how long the game has been played** impacts whether or not the player will come across the bug.

Surveying players in my Discord, as well as watching someone stream the entire game in parts, has lead me to the above conclusion. The players who played the game in shorter bursts of around an hour (give or take) never came across the bug, and nor did the streamer. Conversely, almost all players who had played the game from start to end in one sitting saw it happening at least once (at different places, though). 

One player reported having played the entire game from beginning to end twice, coming across the bug in both those runs (but, again, at different places). When the player went through the game a third and final time, they did so in shorter bursts and the bug never appeared.

I hope this helps in investigating the issue further. :)

**Comment by Tymedust** [comment]
Some new potential breakthroughs have come to light as I've finally managed to capture the issue on camera, courtesy of another let's play I found.

In this video, you can see three examples of the glitch, as well as how it the scene looks without the glitch (sorry about the bad video quality, I had to screen grab the finished video project because export was acting up on my app).

https://github.com/renpy/renpy/assets/123891561/25f63728-1ffe-4c61-be2e-c66b330d37d0

The code for each scene follows below:

Example 1:

```
    e s "That's not what I-"
    $ renpy.pause(0.5, hard=True)
    e a "Stop screwing with me!"
```

Example 2:

```
    rhu "I feel like we've gotten off on the wrong foot here."
    $ renpy.pause(0.5, hard=True)
    hide rh
    show rs 
    rsu "...Or is it \'gotten off on the wrong {i}feet{/i}\'...?"
```

Example 3:

```
    ah "See you later, Isobel!"
    e n "I-"
    stop music fadeout 1
    $ quick_menu = False
    play sound "audio/sfx/teleport.mp3"
    scene hallofrooms with fade
    $ quick_menu = True
    n "What the...?!"
```

What these all have in common is that _the side portrait pops right in when there's some sort of pause, regardless of it being a hard pause as in example 1 or 2, or a fade transition as in example 3._

A **breakthrough** recently came to light as one player suggested that they could replicate one of the examples by _saving at a decision point._ Just prior to example 1, the player comes across the following branching point:

![screenshot0431](https://github.com/renpy/renpy/assets/123891561/5279c5a3-0fdc-446a-b487-9b1479596c5c)

If the player saves here, then the glitch seen in Example 1 will occur, and then more of them will follow. Very interestingly, if the player does _not_ save at this particular point, the glitch will not happen. This is also true if the player saves _before_ or _after_ the decision! _This means that somehow, the act of saving during a choice menu must be connected to the side portrait glitch, right?_ 

Using this method, I was able to accurately re-create the glitch every time. For good measure, I tried to turn the hard pauses into normal, skippable pauses, i.e:

`pause 0.5`

But that didn't make a difference, unfortunately.

I hope that this added information can somehow help further in determining the cause of why this happens and how it can be dealt with. A prior theory I had was that this glitch affects people with low-performing CPUs, but I run a pretty much state-of-the-art rig and I could nonetheless recreate the glitch without fail using the above method.

Thank you for your continued support in helping me out here. :)

**Comment by Org-gith** [comment]
e n "I-"
Not sure but this seems risky, if the narrator is n and some of your side's images are called n. Maybe could rename your narrator to something different like 'nar' to make sure it does not create a conflict somewhere.

**Comment by Janus-Bifrons** [comment]
I know this thread is old and closed, but we've seen this bug many times. We found that changing THIS code just a bit makes it stop
Original
`screen say(who, what):
    style_prefix "say"

    window:
        id "window"

        if who is not None:

            window:
                id "namebox"
                style "namebox"
                text who id "who"

        text what id "what"

    if not renpy.variant("small"):
        add SideImage() xalign 0.0 yalign 1.0`

Altered
`screen say(who, what):
    style_prefix "say"

    window:
        id "window"

        if who is not None:
            if not renpy.variant("small"):
                add SideImage() xalign 0.0 yalign 1.0
            window:
                id "namebox"
                style "namebox"
                text who id "who"

        text what id "what"`
Basically, put the side image code into the "if who is not None:" section, but outside of the window inside of it.
